st3/lsp_utils/server_npm_resource.py

```diff
@@ -85,7 +86,11 @@ def setup(self) -> None:
             return
 
         self._initialized = True
-        self._package_cache_path = os.path.join(sublime.cache_path(), self._package_name)
+        if self._install_in_cache:
+            self._package_cache_path = os.path.join(sublime.cache_path(), self._package_name)
+        else:
+            data_dir = os.path.normpath(os.path.join(sublime.cache_path(), '..'))
+            self._package_cache_path = os.path.join(data_dir, 'Package Storage', self._package_name)
```

Comment:
This doesn't account for the node.js version installed on the machine.

Reply:
No it doesn't, but I like how simple it is without having different folders for different node versions. For this case i would choose simplicity over complexity

Reply:
You can ignore the complexity, but the complexity won't ignore you :)

Some user will eventually pop up and say "yes, I have a portable ST install on my USB stick and my USB stick's flash memory is now dead, because I switch machines all the time and these node modules are constantly being overwritten".

Reply:
I am not saying that we should ignore it, rather wait until we maybe come up with maybe something simpler.

Here is why I think it is complex. If we going to have a folder for each node version than that solution doesn't scale, because there are a lot of node versions out there. (We may end up with at least 6 folders)

We can open an issue for that and discuss the implementation there. (Plus I don't think that node versions are the concern of this PR)

Reply:
> We can open an issue for that and discuss the implementation there. (Plus I don't think that node versions are the concern of this PR)

A fair point. However, given the desire of Rafal to keep everything backwards compatible, you would have to write yet more confusing migration code in another pull request. So I propose to do it right away.

Reply:
I'll add that when I have some free time.

But I suppose it will have to work like so:
 - user installs the package and triggers installation within node v12, for example
 - user updates node to v14 and starts ST - v14 of the server gets installed alongside the v12 server

So in that scenario, the v12 version would not be removed until the user uninstalls the package, even if user is never gonna go back to v12 of node.

Reply:
It would need some refactoring to get the node version before checking if dependencies are installed... Also, that means an extra subprocess start every time the plugin is registered.

Reply:
OK. Doing node version check now before checking if server is installed.

That means that it happens on the main thread but I'll move that to `AbstractPlugin.install_or_update` later for ST4.

I'm only doing the check once per session through a new `NodeVersionResolver` singleton.

Also, I've discovered that in some cases the `ServerNpmResource` would start installing dependencies simultaneously for two or more instances of AbstractPlugin. That happened when ST was started with multiple windows open with each triggering the same LSP plugin. So I've added a `get_server_npm_resource_for_package` utility function that returns an instance of `ServerNpmResource` that is keyed to specific server paths so that multiple instances of the same LSP plugin share the same instance.

Reply:
> That happened when ST was started with multiple windows open with each triggering the same LSP plugin.

Hmm... I don't think it's going to be solvable in LSP lest we end up in callback hell. Oh well. Let's wait till we can upgrade to python 3.8 :-)

Reply:
I think that might an issue specific to ServerNpmResource which starts a new thread for installing node dependencies.

When using `AbstractPlugin.install_or_update` it should run sequentially since LSP will trigger both from the same thread so only first one should trigger installation.

I'm planning to switch to `AbstractPlugin.install_or_update` in next PR. Right now `ServerNpmResource` is triggered from `plugin_loaded`.